### PR TITLE
feat: optional react support

### PR DESCRIPTION
**Pull request by suppayami**
Description
Added basic support for react, using eslint-plugin-react and eslint-plugin-react-hooks.

Linked Issues
Closes https://github.com/antfu/eslint-config/issues/282

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -56,6 +56,8 @@
     "eslint-plugin-n": "^16.3.1",
     "eslint-plugin-no-only-tests": "^3.1.0",
     "eslint-plugin-perfectionist": "^2.4.0",
+    "eslint-plugin-react": "^7.33.2",
+    "eslint-plugin-react-hooks": "^4.6.0",
```
Comment (antfu) [inline package.json:60]: I'd like to keep them as optional peer deps, and ask ppl to explicitly install them when they want to use react. This way we don't introduce those deps to every user of this config.

File: src/configs/react.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+          ecmaFeatures: {
+            jsx: true,
+          },
+        },
+      },
+      name: 'antfu:react:rules',
+      rules: {
+        // recommended rules react
+        ...pluginReact.configs.recommended.rules as any,
+        ...pluginReactHooks.configs.recommended.rules as any,
```
Comment (antfu) [inline src/configs/react.ts:43]: I would like to avoid referencing configs presets, but rather flatten them as inline to have better control.

Comment (suppayami) [reply src/configs/react.ts]: By flattening + inline, do you mean copy the config over?

**Comment by usercao** [comment]
Hi @suppayami , can you add `eslint-plugin-react-refresh`? This is also the most basic configuration item required in `React`.

**Comment by suppayami** [comment]
@usercao I put `eslint-plugin-react-refresh` in, currently `allowConstantExport` is enabled by default, should we only enable it only when certain dev implementation (for example `vite`) exists in deps?

File: package.json (hunk 1 above)
Comment (suppayami) [reply package.json]: I moved them to optional peer deps. Should I leave them in devDeps for dev?

**Comment by usercao** [comment]
@suppayami You're right, I'm sorry I didn't think carefully. Have you been using `CSS-in-JS`? It would be even better if you could refer to the packages listed above and add `CSS-in-JS` support.

**Comment by suppayami** [comment]
@usercao as for css-in-js, I think it will be more appropriate for users to extend the config and put it in themselves, since not everyone use them in React. What do you think of this? @antfu

**Comment by usercao** [comment]
```ts
export default antfu(
  { "vue": true, 
    "typescript": true,
    "styled-components": true,
  },
)
```
I respect your idea, just hope that only one `@antfu/eslint-config` in `devDependencies` is enough.

**Comment by Dimava** [comment]
@suppayami you may use feature flag, same as typescript config has for project (your aware rules won't be enabled if you don't set the tsconfig location)

**Comment by suppayami** [comment]
> @suppayami you may use feature flag, same as typescript config has for project (type aware rules won't be enabled if you don't set the tsconfig location)

Do you mean `allowConstantExport` or css in js? 😂 
As for `allowConstantExport`, I added a list of packages that allow it and auto enable if package(s) exists

**Comment by Dimava** [comment]
@suppayami No I don't, I didn't use React ever 😂
Just noted so you've considered if it's usable in this PR context

**Comment by suppayami** [comment]
@antfu I think feature-wise this is ready to go, please have a review when you have time

**Comment by antfu** [comment]
I updated the PR to not have auto-detection, instead asking users to turn it on manually. Also, it will prompt users to auto-install necessary deps if missing.

**Pull request merged**